rustls/src/server/server_conn.rs

```diff
@@ -540,6 +545,58 @@ impl ServerConfig {
     }
 }
 
+/// A policy describing how invalid an invalid Server Name Indication (SNI) value from a client is handled by the server.
+///
+/// The only valid form of SNI according to relevant RFCs ([RFC6066], [RFC1035]) is
+/// non-IP-address host name, however some misconfigured clients may send a bare IP address, or
+/// another invalid value. Some servers may wish to ignore these invalid values instead of producing
+/// an error.
+///
+/// By default, Rustls will ignore invalid values that are an IP address (the most common misconfiguration)
+/// and error for all other invalid values.
+///
+/// When an SNI value is ignored, Rustls treats the client as if it sent no SNI at all.
+///
+/// [RFC1035]: https://datatracker.ietf.org/doc/html/rfc1035#section-2.3.1
+/// [RFC6066]: https://datatracker.ietf.org/doc/html/rfc6066#section-3
+#[derive(Default, Clone, Copy, PartialEq, Eq, Debug)]
+#[non_exhaustive]
+pub enum InvalidSniPolicy {
+    /// Reject all ClientHello messages that contain an invalid SNI value.
+    RejectAll,
+    /// Ignore an invalid SNI value in ClientHello messages if the value is an IP address.
+    ///
+    /// "Ignoring SNI" means accepting the ClientHello message, but acting as if the client sent no SNI.
+    #[default]
+    IgnoreIpAddresses,
+    /// Ignore all invalid SNI in ClientHello messages.
+    ///
+    /// "Ignoring SNI" means accepting the ClientHello message, but acting as if the client sent no SNI.
+    IgnoreAll,
+}
+
+impl InvalidSniPolicy {
+    /// Check if any potential invalid SNI in ClientHello is acceptable (i.e., ignorable)
+    /// by this policy.
+    ///
+    /// A `false` return value means there is an invalid SNI in ClientHello that is
+    /// rejected by this policy.
+    pub(super) fn is_acceptable(&self, client_hello: &ClientHelloPayload) -> bool {
+        let Some(server_name) = client_hello.server_name.as_ref() else {
+            return true;
+        };
+        match (self, server_name) {
```

Comment:
This is the clippy recommendation: 
```
matches!((self, server_name), (Self::RejectAll, ServerNamePayload::SingleDnsName(_)) | (
                 Self::IgnoreIpAddresses,
                 ServerNamePayload::SingleDnsName(_) | ServerNamePayload::IpAddress,
             ) | (Self::IgnoreAll, _))
```
Which I don't like. Should I apply this clippy recommendation, or work around it?

Reply:
Yuck, I don't like that either :laughing: 

I would be tempted to allow the clippy lint here but we can also wait to see what djc and ctz think. Knowing djc he might have another round of code-golf to suggest that will side-step the concern with an even more improved formulation.

Reply:
@ctz @djc any suggestions?

Reply:
Suggest doing it like this (as per suggestion above):

```rust
match (server_name, self) {
     (Some(ServerNamePayload::SingleDnsName(dns_name)), _) => Some(dns_name.to_lowercase_owned()),
     ...
}
```